### PR TITLE
feat: add PAT monitoring functions

### PR DESCRIPTION
**Pull request by rickstaa**
This commit adds two monitoring functions that can be used to check
whether the PATs are functioning correctly:
 - status/up: Returns whether the PATs are rate limited.
 - status/pat-info: Returns information about the PATs.

@anuraghazra I created this PR for you to review the two PAT monitoring functions I added to the https://github.com/anuraghazra/github-readme-stats/tree/monitoring branch. We can, however, keep these cloud functions on the monitoring branch since they are not meant to be used by the users. 

The `up` function is meant to be run in dashboard software like [Grafana](https://grafana.com/) to see how many times in a given day the PATs were limited. The `pat-info` function can check if PATs are expired. I added caching headers to these functions to prevent abuse (i.e. 1 req per 10 min).

### See in action

**Up:** https://github-readme-stats-git-monitoring-github-readme-stats-team.vercel.app/api/status/up.
**PAT-info:** https://github-readme-stats-git-monitoring-github-readme-stats-team.vercel.app/api/status/pat-info.

### Dynamic badge

When you supply the `?type=shields` argument to the Up function, it will return a [dynamic shields.io JSON object](https://shields.io/endpoint). This object can be used to generate a [shields.io badge](https://shields.io/endpoint) that shows whether the Public Vercel instance is rate limited. We can change the badge label and message to our liking. You can also retrieve the result as a JSON object using the `type=json` argument.

Custom badge

```md
<img alt="Custom badge" src="https://img.shields.io/endpoint?url=https%3A%2F%2Fgithub-readme-stats-git-monitoring-github-readme-stats-team.vercel.app%2Fapi%2Fstatus%2Fup%3Ftype%3Dshields">
```

### Result

Not sure if the `pa-info` function is working correctly. It worked locally and for my own instance (see https://github-readme-stats-git-monitoring-rickstaa.vercel.app/api/status/pat-info), but for the public instance, it looks like we only have one valid PAT. I think the function puts `rate_limited` PATs in the expired field. I couldn't find information about the exact behaviour (i.e. error thrown) when a PAT token is limited since I didn't have a rate-limited token.

image

### CODE REVIEW

File: api/status/pat-info.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/**
+ * @file Contains a simple cloud function that can be used to check which PATs are no
+ * longer working. It returns a list of valid PATs, expired PATs and PATs with errors.
+ *
+ * @description This function is currently rate limited to 1 request per day.
+ */
+
+import { logger, request } from "../../src/common/utils.js";
+
+export const RATE_LIMIT_SECONDS = 60 * 60 * 24; // 1 request per day.
```
Comment (rickstaa) [inline api/status/pat-info.js:10]: We can change this limit to our likings.

**Comment by Zo-Bro-23** [comment]
Any particular reason for not merging? This seems pretty useful to me for monitoring uptime.

**Comment by rickstaa** [comment]
> Any particular reason for not merging? This seems pretty useful to me for monitoring uptime.
> 
> Edit: Your domain works well, but wouldn't it be better if all features were included in `master`?

There is no reason other than this being an OS project and I want to give @anuraghazra all the time he needs to review it before we merge it. 👍

**Comment by Zo-Bro-23** [comment]
> > Any particular reason for not merging? This seems pretty useful to me for monitoring uptime.
> > Edit: Your domain works well, but wouldn't it be better if all features were included in `master`?
> 
> There is no reason other than this being an OS project and I want to give @anuraghazra all the time he needs to review it before we merge it. +1

Right. There's a small chance @anuraghazra might have forgotten about this though, since the last comment was on Oct 2022 😅

**Comment by anuraghazra** [comment]
Yeah I forgot, I did the review and made few changes. 

I'll take a final look today and merge it.

**Comment by anuraghazra** [comment]
Tests are failing can you check?

**Comment by rickstaa** [comment]
> Tests are failing can you check?

Of course! Your new code stringifies the response where my tests expected a JS object. Fixed 👍🏻.

**Comment by anuraghazra** [comment]
shield badge is failing 🤔 
 image

File: api/status/pat-info.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      } else {
+        details[pat] = {
+          status: "valid",
+          remaining: response.data.data.rateLimit.remaining,
+        };
+      }
+    } catch (err) {
+      // Store the PAT if it is expired.
+      const errorMessage = err.response?.data?.message;
+      if (errorMessage?.toLowerCase() === "bad credentials") {
```
Comment (anuraghazra) [inline api/status/pat-info.js:80]: Should probably also add the new error which we fixed in retryer. 

`isAccountSuspended`

File: api/status/pat-info.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+/**
+ * Cloud function that returns information about the used PATs.
+ */
+export default async (_, res) => {
+  res.setHeader("Content-Type", "application/json");
+  try {
+    // Add header to prevent abuse.
+    const pATsInfo = await getPATInfo(uptimeFetcher, {});
```
Comment (anuraghazra) [inline api/status/pat-info.js:110]:
```suggestion
    const PATsInfo = await getPATInfo(uptimeFetcher, {});
```

**Review by anuraghazra — APPROVED** [review]
Few minor comments, you can merge after resolving those. Thanks!

**Comment by rickstaa** [comment]
> shield badge is failing thinking image

I quickly checked why the shields were not working. It looks like that although the PR works locally, Vercel gives a 403 error:
- Local endpoint: http://localhost:3000/api/status/up?type=shields
- Remote endpoint: https://github-readme-stats-git-monitoring-github-readme-stats-team.vercel.app/api/status/up?type=shields 

I use https://shields.io/endpoint for the badge. I am not a member of the Vercel GRS team, so I can not see what goes wrong. 😅

**Pull request merged**